lib/surface/compiler/eex_engine.ex

```diff
@@ -102,6 +102,45 @@ defmodule Surface.Compiler.EExEngine do
     |> maybe_print_expression(comprehension)
   end
 
+  defp to_expression(
+         %AST.For{
+           generator: %AST.AttributeExpr{value: generator},
+           children: children,
+           else: else_children
+         } = comprehension,
+         buffer,
+         state
+       ) do
+    for_buffer =
+      handle_nested_block(children, buffer, %{
+        state
+        | depth: state.depth + 1,
+          context: [:for | state.context]
+      })
+
+    else_buffer =
+      handle_nested_block(else_children, buffer, %{
+        state
+        | depth: state.depth + 1,
+          context: [:for | state.context]
+      })
+
+    generator_expr = generator ++ [[do: for_buffer]]
+    for_expr = {:for, [generated: true], generator_expr}
+
+    case_expr =
+      quote generated: true do
+        unquote(for_expr)
+        |> case do
+          [] -> unquote(else_buffer)
+          result -> result
+        end
+      end
```

Comment:
@lnr0626 @Malian `for_else` works fine when passing regular html as children, but breaks when passing other components with error (both as child of the `for` or the `else`):
> ** (ArgumentError) cannot convert component Surface.Constructs.ForTest.SomeComponent with id nil to HTML.

It happens in this highlighted piece of code, any hints?

Reply:
@miguel-s that's bad news. Does the `<#for>` works as expected with components if you don't use `<#else>`? If so, that means we can't use the pipe :(

Not sure how we can do this without creating a variable, which will mess with change tracking. Maybe we just can't. I'll try to come up with something but there's a big chance we'll have to let the `for...else` go.

Reply:
Another solution would be to allow `for...else` only when it's using a single generator and no filters. Although it's not an ideal solution, using multiple generators and filters in templates is quite rare so maybe it's worth it. 

From the implementation PoV, all we'd have to do is to generate two pieces of code. For instance, the following code:

```
<#for {i <- whatever}>
  CONTENT
<#else>
  ELSE CONTENT
</#for>
```
would be equivalent to:

```
<#for {i <- whatever}>
  CONTENT
</#for>
<#if {whatever != []}>
  ELSE CONTENT
</#if>
```

The funny thing is that the above code looks exactly how most of the developers would handle this situation when they want to show a fallback-content if the list is empty.

This solution comes with a tiny performance issue if the user decides to use a more expensive expression. For instance:

```
<#for {i <- Enum.map(@items, &some_expensive_calculation/1)}>
  ...
<#else>
  ...
</#for>
```

In that case, `Enum.map/2` would always run twice for each rendering.

We could also try to generate something like:

``` 
case Enum.map(@items, &some_expensive_calculation/1)} do
  result ->
    for {i <- result}>
      ...
    end
end
```

But I'm not sure it would compromise change tracking. I'll try to create some tests so we can evaluate that. Unless, of course, we just decide to remove `for...else` from our list :)

My opinion is that, since we know when there's an `<#else>` and since we can pattern match against the expression to make sure we only accept a single generator so we can raise a nice error message when the expression does not conform to that, it seems to be worth it to give it a try.

Any other ideas?

Reply:
@miguel-s I played around with `for` using some vanilla EEX and I couldn't see any difference regarding change tracking when wrapping it in a `case`. So, in case we decide to keep `for...else` restricting its use only when passing a single generator without filters, the following should work just fine:

```
<#for {i <- whatever}>
  CONTENT
<#else>
  ELSE CONTENT
</#for>
```

becomes:

```
case whatever do
  [] ->
    ELSE CONTENT
  result ->
    for {i <- result}>
      CONTENT
    end
end
```

Reply:
What a bad news :cry: 

> Although it's not an ideal solution, using multiple generators and filters in templates is quite rare so maybe it's worth it.

I would say that using multiple generators, filters __and else__ in templates is quite rare. Even with the simple `for/else` use case I believe we create a lot of value for developers.

Reply:
@msaraiva @Malian I got it to work with the compiler version (nesting `AST.For` inside `AST.If`), but have not been able to get it to work in the eex_engine version.

These are the issues I'm running into:

#### Attempt 1

```elixir
[{:<-, _, [{_, _, _}, value]}] = generator # `value` is the generator e.g. `["apples", "bananas", "oranges"]`

if_expr =
  quote generated: true do
    if unquote(value) == [] do
      unquote(else_buffer)
    else
      unquote(for_expr)
    end
  end
```

Everything inside `if` runs fine, but `else` still raises:
> ** (ArgumentError) cannot convert component Surface.Constructs.ForTest.SomeComponent with id nil to HTML.

#### Attempt 2

```elixir
[{:<-, _, [{var, _, _}, value]}] = generator # `var` is the name of the defined variable e.g. `:fruit` ; `value` is the generator e.g. `["apples", "bananas", "oranges"]`

if_expr =
  quote generated: true do
    if unquote(value) == [] do
      unquote(else_buffer)
    else
      for unquote(var) <- unquote(value) do
        unquote(for_buffer)
      end
    end
  end
```

This raises:
> ** (CompileError) test/constructs/for_test.exs:190: undefined function fruit/0

The `fruit` variable comes from the generator in the test `<#for each={fruit <- ["apples", "bananas", "oranges"]}>`

It feels like I'm getting very close, but I don't know how to solve this last error 😢

Reply:
I believe it is
```elixir
[{:<-, _, [{_, _, _} = var, value]}] = generator # `var` is the name of the defined variable e.g. `:fruit` ; `value` is the generator e.g. `["apples", "bananas", "oranges"]`

if_expr =
  quote generated: true do
    if unquote(value) == [] do
      unquote(else_buffer)
    else
      for unquote(var) <- unquote(value) do
        unquote(for_buffer)
      end
    end
  end
```
but from my point of view it is exactly the same as
```elixir
[{:<-, _, [{_, _, _}, value]}] = generator # `value` is the generator e.g. `["apples", "bananas", "oranges"]`

if_expr =
  quote generated: true do
    if unquote(value) == [] do
      unquote(else_buffer)
    else
      unquote(for_expr)
    end
  end
```
or
```elixir
    case_expr =
      quote generated: true do
        case unquote(value) do
          [] -> unquote(else_buffer)
          result -> unquote(for_expr)
        end
      end
```
I am not expert enough with macros to go further.

Reply:
@Malian you are right, your solution (I must be blind to not have seen it) fixed the CompileError, but we're back to:

> ** (ArgumentError) cannot convert component Surface.Constructs.ForTest.SomeComponent with id nil to HTML.

Reply:
> I got it to work with the compiler version (nesting `AST.For` inside `AST.If`), but have not been able to get it to work in the eex_engine version.

@miguel-s implementing that in the eex_engine will be much harder as you can't generate the whole quoted expression like you'd normally do when creating a macro. You'll have to emit each piece of code according to what EEX expects. So I'd just do that at the compiler level using `AST.If` + `AST.For` as they should be properly handled by the engine, emitting the correct bits calling the underlying engine's `handle_*` callbacks. 

I'm glad you tried to use it with components and found that issue. We need to make sure all constructs follow that same rule, otherwise, they would work nice with HTML/text content but fail when wrapping a live component. We should add a test for that scenario to every constructwe have.

Reply:
Pushed the compiler version and added some tests to make sure `#if` accepts components